examples/rp2/pio_quadrature_encoder.py

```diff
@@ -0,0 +1,100 @@
+"""
+This implementation offloads the quadrature sensor continuous pooling to the StateMachine pio_quadrature,
+pio_quadrature will detect state changes and push the values then raises an interrupt that is handled by
+encoder_state_changed_irq_handler
+
+encoder_state_changed_irq_handler was inspired from 
+https://electronics.stackexchange.com/questions/360637/quadrature-encoder-most-efficient-software-implementation
+
+"""
+
+from machine import Pin
+import rp2
+from rp2 import PIO, StateMachine
+import utime
+import array
+
+
+state_look_up_table = array.array("b", [
+        #Direction = 1
+        0, # 00 to 00
+       -1, # 00 to 01
+       +1, # 00 to 10
+       +2, # 00 to 11
+
+       +1, # 01 to 00
+        0, # 01 to 01
+       +2, # 01 to 10
+       -1, # 01 to 11
+
+       -1, # 10 to 00
+       +2, # 10 to 01
+        0, # 10 to 10
+       +1, # 10 to 11
+
+       +2, # 11 to 00
+       +1, # 11 to 01
+       -1, # 11 to 10
+        0, # 11 to 11
+
+        #Direction = 0
+        0, # 00 to 00
+       -1, # 00 to 01
+       +1, # 00 to 10
+       -2, # 00 to 11
+
+       +1, # 01 to 00
+        0, # 01 to 01
+       -2, # 01 to 10
+       -1, # 01 to 11
+
+       -1, # 10 to 00
+       -2, # 10 to 01
+        0, # 10 to 10
+       +1, # 10 to 11
+
+       -2, # 11 to 00
+       +1, # 11 to 01
+       -1, # 11 to 10
+        0, # 11 to 11
+    ])
+
+
+counter = 0
+direction = 0
+lut_index = 0
+
+def encoder_state_changed_irq_handler(sm):
+    global counter, direction, lut_index
+    lut_index |= sm.get() & 3
```

Comment:
I think this would be better as a loop, to cater for the case where another state change occurs while the ISR is running, e.g.
```python
def encoder_state_changed_irq_handler(sm):
    global counter, direction, lut_index
    while sm.rx_fifo():
        lut_index |= sm.get() & 3
        counter += state_look_up_table[lut_index]
        if state_look_up_table[lut_index] != 0:
            direction = 1 if (state_look_up_table[lut_index] > 0) else 0
        lut_index = ((lut_index << 2) & 0b1100) | (direction << 4)
```

Reply:
That improves the speed indeed.